File: py/tests/test_registry.py
import pytest

from wlearn.registry import register, load, get_registry
from wlearn.bundle import encode_bundle
from wlearn.errors import RegistryError


def _make_bundle(type_id):
    return encode_bundle(
        {'typeId': type_id},
        [{'id': 'model', 'data': b'\x01\x02\x03'}],
    )


def test_register_and_load():
    register('test.reg.basic@1', lambda m, t, b: {'loaded': True, 'typeId': m['typeId']})
    bundle = _make_bundle('test.reg.basic@1')
    result = load(bundle)
    assert result['loaded'] is True
    assert result['typeId'] == 'test.reg.basic@1'


def test_register_invalid_type_id():
    with pytest.raises(RegistryError, match='must contain "@"'):
        register('no-version', lambda m, t, b: None)


def test_register_invalid_loader():
    with pytest.raises(RegistryError, match='callable'):
        register('test.bad@1', 'not a function')


def test_load_missing_type_id():
    bundle = _make_bundle('test.reg.nonexistent@99')
    with pytest.raises(RegistryError, match='No loader registered'):
        load(bundle)


def test_load_error_lists_available():
    register('test.reg.available@1', lambda m, t, b: None)
    bundle = _make_bundle('test.reg.missing@1')
    with pytest.raises(RegistryError, match='test.reg.available@1'):
        load(bundle)


def test_get_registry_returns_copy():
    register('test.reg.copy@1', lambda m, t, b: None)
    reg = get_registry()
    assert 'test.reg.copy@1' in reg
    # mutating the copy should not affect the real registry
    reg.clear()
    assert 'test.reg.copy@1' in get_registry()

File: packages/core/test/test-registry.js
const { describe, it, beforeEach } = require('node:test')
const assert = require('node:assert/strict')
const { register, load, loadSync, getRegistry } = require('../src/registry.js')
const { encodeBundle } = require('../src/bundle.js')
const { RegistryError } = require('../src/errors.js')

// Helper: create a minimal bundle
function makeBundle(typeId, data = new Uint8Array([1])) {
  return encodeBundle({ typeId }, [{ id: 'model', data }])
}

describe('register', () => {
  it('validates typeId contains @', () => {
    assert.throws(
      () => register('bad-type-id', () => {}),
      RegistryError
    )
  })

  it('validates loaderFn is a function', () => {
    assert.throws(
      () => register('wlearn.test@1', 'not a function'),
      RegistryError
    )
  })

  it('registers a loader', () => {
    register('wlearn.test.register@1', () => 'ok')
    const reg = getRegistry()
    assert(reg.has('wlearn.test.register@1'))
  })
})

describe('load (async)', () => {
  it('dispatches to sync loader and returns Promise', async () => {
    const mockEstimator = { isFitted: true }
    register('wlearn.test.sync-loader@1', () => mockEstimator)

    const bytes = makeBundle('wlearn.test.sync-loader@1')
    const result = load(bytes)
    assert(result instanceof Promise)
    assert.strictEqual(await result, mockEstimator)
  })

  it('dispatches to async loader', async () => {
    const mockEstimator = { isFitted: true }
    register('wlearn.test.async-loader@1', async () => mockEstimator)

    const bytes = makeBundle('wlearn.test.async-loader@1')
    const result = await load(bytes)
    assert.strictEqual(result, mockEstimator)
  })

  it('throws RegistryError for missing loader', async () => {
    const bytes = makeBundle('wlearn.unknown@1')
    await assert.rejects(() => load(bytes), (err) => {
      assert(err instanceof RegistryError)
      assert(err.message.includes('wlearn.unknown@1'))
      assert(err.message.includes('No loader registered'))
      return true
    })
  })

  it('error includes available loaders', async () => {
    register('wlearn.test.listed@1', () => {})
    const bytes = makeBundle('wlearn.not-here@1')
    await assert.rejects(() => load(bytes), (err) => {
      assert(err.message.includes('wlearn.test.listed@1'))
      return true
    })
  })
})

describe('loadSync', () => {
  it('works with sync loader', () => {
    const mockEstimator = { type: 'sync' }
    register('wlearn.test.sync-only@1', () => mockEstimator)
    const bytes = makeBundle('wlearn.test.sync-only@1')
    assert.strictEqual(loadSync(bytes), mockEstimator)
  })

  it('throws if loader returns Promise', () => {
    register('wlearn.test.async-only@1', async () => ({ type: 'async' }))
    const bytes = makeBundle('wlearn.test.async-only@1')
    assert.throws(() => loadSync(bytes), RegistryError)
  })

  it('throws for missing loader', () => {
    const bytes = makeBundle('wlearn.no-such@1')
    assert.throws(() => loadSync(bytes), RegistryError)
  })
})

describe('getRegistry', () => {
  it('returns a copy', () => {
    register('wlearn.test.copy@1', () => {})
    const reg = getRegistry()
    reg.delete('wlearn.test.copy@1')
    // Internal registry unaffected
    const reg2 = getRegistry()
    assert(reg2.has('wlearn.test.copy@1'))
  })
})
